File: app/src/main/java/org/stepic/droid/storage/dao/CourseDaoImpl.java
package org.stepic.droid.storage.dao;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.jetbrains.annotations.Nullable;
import org.stepic.droid.model.CachedVideo;
import org.stepic.droid.model.Course;
import org.stepic.droid.model.Video;
import org.stepic.droid.model.VideoUrl;
import org.stepic.droid.storage.structure.DbStructureEnrolledAndFeaturedCourses;
import org.stepic.droid.storage.structure.DbStructureCachedVideo;
import org.stepic.droid.util.DbParseHelper;
import org.stepic.droid.util.VideoCourseHelper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class CourseDaoImpl extends DaoBase<Course> {
    private final IDao<CachedVideo> cachedVideoDao;
    private String tableName;

    @Inject
    public CourseDaoImpl(SQLiteDatabase openHelper, IDao<CachedVideo> cachedVideoDao, String tableName) {
        super(openHelper);
        this.cachedVideoDao = cachedVideoDao;
        this.tableName = tableName;
    }

    @Override
    public Course parsePersistentObject(Cursor cursor) {
        Course course = new Course();

        int indexId = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.COURSE_ID);
        int indexSummary = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.SUMMARY);
        int indexCover = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.COVER_LINK);
        int indexIntro = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.INTRO_LINK_VIMEO);
        int indexTitle = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.TITLE);
        int indexLanguage = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.LANGUAGE);
        int indexBeginDateSource = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.BEGIN_DATE_SOURCE);
        int indexBeginDate = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.BEGIN_DATE);
        int indexEndDate = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.END_DATE);
        int indexLastDeadline = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.LAST_DEADLINE);
        int indexDescription = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.DESCRIPTION);
        int indexInstructors = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.INSTRUCTORS);
        int indexRequirements = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.REQUIREMENTS);
        int indexEnrollment = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.ENROLLMENT);
        int indexSection = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.SECTIONS);
        int indexWorkload = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.WORKLOAD);
        int indexCourseFormat = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.COURSE_FORMAT);
        int indexTargetAudience = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.TARGET_AUDIENCE);
        int indexCertificate = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.CERTIFICATE);
        int indexIntroVideoId = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.INTRO_VIDEO_ID);
        int indexSlug = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.SLUG);
        int indexScheduleLink = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.SCHEDULE_LINK);
        int indexScheduleLongLink = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.SCHEDULE_LONG_LINK);
        int indexLastStepId = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.LAST_STEP_ID);
        int indexIsActive = cursor.getColumnIndex(DbStructureEnrolledAndFeaturedCourses.Column.IS_ACTIVE);

        course.setLastStepId(cursor.getString(indexLastStepId));
        course.setCertificate(cursor.getString(indexCertificate));
        course.setWorkload(cursor.getString(indexWorkload));
        course.setCourse_format(cursor.getString(indexCourseFormat));
        course.setTarget_audience(cursor.getString(indexTargetAudience));

        course.setId(cursor.getLong(indexId));
        course.setSummary(cursor.getString(indexSummary));
        course.setCover(cursor.getString(indexCover));
        course.setIntro(cursor.getString(indexIntro));
        course.setTitle(cursor.getString(indexTitle));
        course.setLanguage(cursor.getString(indexLanguage));
        course.setBegin_date_source(cursor.getString(indexBeginDateSource));
        course.setLast_deadline(cursor.getString(indexLastDeadline));
        course.setDescription(cursor.getString(indexDescription));
        course.setInstructors(DbParseHelper.parseStringToLongArray(cursor.getString(indexInstructors)));
        course.setRequirements(cursor.getString(indexRequirements));
        course.setEnrollment(cursor.getInt(indexEnrollment));
        course.setSections(DbParseHelper.parseStringToLongArray(cursor.getString(indexSection)));
        course.setIntro_video_id(cursor.getLong(indexIntroVideoId));
        course.setSlug(cursor.getString(indexSlug));
        course.setSchedule_link(cursor.getString(indexScheduleLink));
        course.setSchedule_long_link(cursor.getString(indexScheduleLongLink));
        course.setBegin_date(cursor.getString(indexBeginDate));
        course.setEnd_date(cursor.getString(indexEndDate));

        boolean isActive = true;
        try {
            isActive = cursor.getInt(indexIsActive) > 0;
        } catch (Exception exception) {
            //it can be null before migration --> default active
        }
        course.setIs_active(isActive);

        return course;
    }

    @Override
    public ContentValues getContentValues(Course course) {
        ContentValues values = new ContentValues();

        values.put(DbStructureEnrolledAndFeaturedCourses.Column.LAST_STEP_ID, course.getLastStepId());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.COURSE_ID, course.getCourseId());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.SUMMARY, course.getSummary());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.COVER_LINK, course.getCover());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.INTRO_LINK_VIMEO, course.getIntro());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.TITLE, course.getTitle());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.LANGUAGE, course.getLanguage());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.BEGIN_DATE_SOURCE, course.getBegin_date_source());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.LAST_DEADLINE, course.getLast_deadline());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.DESCRIPTION, course.getDescription());

        String instructorsParsed = DbParseHelper.parseLongArrayToString(course.getInstructors());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.INSTRUCTORS, instructorsParsed);

        values.put(DbStructureEnrolledAndFeaturedCourses.Column.REQUIREMENTS, course.getRequirements());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.ENROLLMENT, course.getEnrollment());

        String sectionsParsed = DbParseHelper.parseLongArrayToString(course.getSections());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.SECTIONS, sectionsParsed);

        values.put(DbStructureEnrolledAndFeaturedCourses.Column.WORKLOAD, course.getWorkload());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.COURSE_FORMAT, course.getCourse_format());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.TARGET_AUDIENCE, course.getTarget_audience());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.CERTIFICATE, course.getCertificate());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.SLUG, course.getSlug());

        values.put(DbStructureEnrolledAndFeaturedCourses.Column.SCHEDULE_LINK, course.getSchedule_link());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.SCHEDULE_LONG_LINK, course.getSchedule_long_link());

        values.put(DbStructureEnrolledAndFeaturedCourses.Column.BEGIN_DATE, course.getBegin_date());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.END_DATE, course.getEnd_date());
        values.put(DbStructureEnrolledAndFeaturedCourses.Column.IS_ACTIVE, course.is_active());

        Video video = course.getIntro_video();
        if (video != null) {
            values.put(DbStructureEnrolledAndFeaturedCourses.Column.INTRO_VIDEO_ID, video.getId());
        }
        return values;
    }

    @Override
    public String getDbName() {
        return tableName;
    }

    @Override
    public String getDefaultPrimaryColumn() {
        return DbStructureEnrolledAndFeaturedCourses.Column.COURSE_ID;
    }

    @Override
    public String getDefaultPrimaryValue(Course persistentObject) {
        return persistentObject.getCourseId() + "";
    }

    @Nullable
    @Override
    public Course get(String whereColumn, String whereValue) {
        Course course = super.get(whereColumn, whereValue);
        addInnerObjects(course);
        return course;
    }

    @Override
    protected List<Course> getAllWithQuery(String query, String[] whereArgs) {
        List<Course> courseList = super.getAllWithQuery(query, whereArgs);
        for (Course course : courseList) {
            addInnerObjects(course);
        }
        return courseList;
    }

    private void addInnerObjects(Course course) {
        if (course == null) return;
        CachedVideo video = cachedVideoDao.get(DbStructureCachedVideo.Column.VIDEO_ID, course.getIntro_video_id() + "");
        if (video != null) {
            course.setIntro_video(transformCachedVideoToRealVideo(video));
        }
    }

    @Override
    public void insertOrUpdate(Course course) {
        super.insertOrUpdate(course);
        if (course != null && course.getIntro_video() != null) {
            Video video = course.getIntro_video();
            CachedVideo cachedVideo = VideoCourseHelper.INSTANCE.transformToCachedVideo(video); //it is cached, but not stored video.
            cachedVideoDao.insertOrUpdate(cachedVideo);
        }
    }

    //// FIXME: 17.02.16 refactor this hack
    @Nullable
    private Video transformCachedVideoToRealVideo(CachedVideo cachedVideo) {
        Video realVideo = null;
        if (cachedVideo != null) {
            realVideo = new Video();
            realVideo.setId(cachedVideo.getVideoId());
            realVideo.setThumbnail(cachedVideo.getThumbnail());
            VideoUrl videoUrl = new VideoUrl();
            videoUrl.setQuality(cachedVideo.getQuality());
            videoUrl.setUrl(cachedVideo.getUrl());

            List<VideoUrl> list = new ArrayList<>();
            list.add(videoUrl);
            realVideo.setUrls(list);
        }
        return realVideo;
    }
}

File: app/src/main/java/org/stepic/droid/storage/dao/CourseDaoImpl.kt
package org.stepic.droid.storage.dao

import android.content.ContentValues
import android.database.Cursor
import com.google.gson.Gson
import org.stepic.droid.storage.operations.DatabaseOperations
import org.stepic.droid.storage.structure.DbStructureCourse
import org.stepic.droid.util.DbParseHelper
import org.stepic.droid.util.getBoolean
import org.stepic.droid.util.getDouble
import org.stepic.droid.util.getLong
import org.stepic.droid.util.getString
import org.stepic.droid.util.toObject
import org.stepik.android.cache.video.dao.VideoDao
import org.stepik.android.model.Course
import org.stepik.android.model.Video
import javax.inject.Inject

class CourseDaoImpl
@Inject
constructor(
    databaseOperations: DatabaseOperations,
    private val videoDao: VideoDao,
    private val gson: Gson
) : DaoBase<Course>(databaseOperations) {

    public override fun getDbName() = DbStructureCourse.TABLE_NAME
    public override fun getDefaultPrimaryColumn() = DbStructureCourse.Columns.ID

    public override fun getDefaultPrimaryValue(persistentObject: Course) = persistentObject.id.toString()

    public override fun parsePersistentObject(cursor: Cursor): Course =
        Course(
            id = cursor.getLong(DbStructureCourse.Columns.ID),
            title = cursor.getString(DbStructureCourse.Columns.TITLE),
            description = cursor.getString(DbStructureCourse.Columns.DESCRIPTION),
            cover = cursor.getString(DbStructureCourse.Columns.COVER),
            certificate = cursor.getString(DbStructureCourse.Columns.CERTIFICATE),
            requirements = cursor.getString(DbStructureCourse.Columns.REQUIREMENTS),
            summary = cursor.getString(DbStructureCourse.Columns.SUMMARY),
            workload = cursor.getString(DbStructureCourse.Columns.WORKLOAD),
            intro = cursor.getString(DbStructureCourse.Columns.INTRO),
            introVideo = Video(id = cursor.getLong(DbStructureCourse.Columns.INTRO_VIDEO_ID)),
            language = cursor.getString(DbStructureCourse.Columns.LANGUAGE),
            authors = DbParseHelper.parseStringToLongList(cursor.getString(DbStructureCourse.Columns.AUTHORS)),
            instructors = DbParseHelper.parseStringToLongList(cursor.getString(DbStructureCourse.Columns.INSTRUCTORS)),
            sections = DbParseHelper.parseStringToLongList(cursor.getString(DbStructureCourse.Columns.SECTIONS)),
            courseFormat = cursor.getString(DbStructureCourse.Columns.COURSE_FORMAT),
            targetAudience = cursor.getString(DbStructureCourse.Columns.TARGET_AUDIENCE),
            certificateFooter = cursor.getString(DbStructureCourse.Columns.CERTIFICATE_FOOTER),
            certificateCoverOrg = cursor.getString(DbStructureCourse.Columns.CERTIFICATE_COVER_ORG),
            totalUnits = cursor.getLong(DbStructureCourse.Columns.TOTAL_UNITS),
            enrollment = cursor.getLong(DbStructureCourse.Columns.ENROLLMENT),
            progress = cursor.getString(DbStructureCourse.Columns.PROGRESS),
            owner = cursor.getLong(DbStructureCourse.Columns.OWNER),
            readiness = cursor.getDouble(DbStructureCourse.Columns.READINESS),
            isContest = cursor.getBoolean(DbStructureCourse.Columns.IS_CONTEST),
            isFeatured = cursor.getBoolean(DbStructureCourse.Columns.IS_FEATURED),
            isActive = cursor.getBoolean(DbStructureCourse.Columns.IS_ACTIVE),
            isPublic = cursor.getBoolean(DbStructureCourse.Columns.IS_PUBLIC),
            isArchived = cursor.getBoolean(DbStructureCourse.Columns.IS_ARCHIVED),
            isFavorite = cursor.getBoolean(DbStructureCourse.Columns.IS_FAVORITE),
            certificateDistinctionThreshold = cursor.getLong(DbStructureCourse.Columns.CERTIFICATE_DISTINCTION_THRESHOLD),
            certificateRegularThreshold = cursor.getLong(DbStructureCourse.Columns.CERTIFICATE_REGULAR_THRESHOLD),
            certificateLink = cursor.getString(DbStructureCourse.Columns.CERTIFICATE_LINK),
            isCertificateAutoIssued = cursor.getBoolean(DbStructureCourse.Columns.IS_CERTIFICATE_AUTO_ISSUED),
            isCertificateIssued = cursor.getBoolean(DbStructureCourse.Columns.IS_CERTIFICATE_ISSUED),
            lastDeadline = cursor.getString(DbStructureCourse.Columns.LAST_DEADLINE),
            beginDate = cursor.getString(DbStructureCourse.Columns.BEGIN_DATE),
            endDate = cursor.getString(DbStructureCourse.Columns.END_DATE),
            slug = cursor.getString(DbStructureCourse.Columns.SLUG),

            scheduleLink = cursor.getString(DbStructureCourse.Columns.SCHEDULE_LINK),
            scheduleLongLink = cursor.getString(DbStructureCourse.Columns.SCHEDULE_LONG_LINK),
            scheduleType = cursor.getString(DbStructureCourse.Columns.SCHEDULE_TYPE),

            lastStepId = cursor.getString(DbStructureCourse.Columns.LAST_STEP),
            learnersCount = cursor.getLong(DbStructureCourse.Columns.LEARNERS_COUNT),
            reviewSummary = cursor.getLong(DbStructureCourse.Columns.REVIEW_SUMMARY),
            timeToComplete = cursor.getLong(DbStructureCourse.Columns.TIME_TO_COMPLETE),
            courseOptions = cursor.getString(DbStructureCourse.Columns.OPTIONS)?.toObject(gson),

            isPaid = cursor.getBoolean(DbStructureCourse.Columns.IS_PAID),
            price = cursor.getString(DbStructureCourse.Columns.PRICE),
            currencyCode = cursor.getString(DbStructureCourse.Columns.CURRENCY_CODE),
            displayPrice = cursor.getString(DbStructureCourse.Columns.DISPLAY_PRICE),
            priceTier = cursor.getString(DbStructureCourse.Columns.PRICE_TIER)
        )

    public override fun getContentValues(course: Course): ContentValues {
        val values = ContentValues()

        values.put(DbStructureCourse.Columns.ID, course.id)
        values.put(DbStructureCourse.Columns.TITLE, course.title)
        values.put(DbStructureCourse.Columns.DESCRIPTION, course.description)
        values.put(DbStructureCourse.Columns.COVER, course.cover)
        values.put(DbStructureCourse.Columns.CERTIFICATE, course.certificate)
        values.put(DbStructureCourse.Columns.REQUIREMENTS, course.requirements)
        values.put(DbStructureCourse.Columns.SUMMARY, course.summary)
        values.put(DbStructureCourse.Columns.WORKLOAD, course.workload)
        values.put(DbStructureCourse.Columns.INTRO, course.intro)
        values.put(DbStructureCourse.Columns.INTRO_VIDEO_ID, course.introVideo?.id ?: -1) // todo add complete course entity and remove this hack
        values.put(DbStructureCourse.Columns.LANGUAGE, course.language)
        values.put(DbStructureCourse.Columns.AUTHORS, DbParseHelper.parseLongListToString(course.authors))
        values.put(DbStructureCourse.Columns.INSTRUCTORS, DbParseHelper.parseLongListToString(course.instructors))
        values.put(DbStructureCourse.Columns.SECTIONS, DbParseHelper.parseLongListToString(course.sections))
        values.put(DbStructureCourse.Columns.COURSE_FORMAT, course.courseFormat)
        values.put(DbStructureCourse.Columns.TARGET_AUDIENCE, course.targetAudience)
        values.put(DbStructureCourse.Columns.CERTIFICATE_FOOTER, course.certificateFooter)
        values.put(DbStructureCourse.Columns.CERTIFICATE_COVER_ORG, course.certificateCoverOrg)
        values.put(DbStructureCourse.Columns.TOTAL_UNITS, course.totalUnits)
        values.put(DbStructureCourse.Columns.ENROLLMENT, course.enrollment)
        values.put(DbStructureCourse.Columns.PROGRESS, course.progress)
        values.put(DbStructureCourse.Columns.OWNER, course.owner)
        values.put(DbStructureCourse.Columns.READINESS, course.readiness)
        values.put(DbStructureCourse.Columns.IS_CONTEST, course.isContest)
        values.put(DbStructureCourse.Columns.IS_FEATURED, course.isFeatured)
        values.put(DbStructureCourse.Columns.IS_ACTIVE, course.isActive)
        values.put(DbStructureCourse.Columns.IS_PUBLIC, course.isPublic)
        values.put(DbStructureCourse.Columns.IS_ARCHIVED, course.isArchived)
        values.put(DbStructureCourse.Columns.IS_FAVORITE, course.isFavorite)
        values.put(DbStructureCourse.Columns.CERTIFICATE_DISTINCTION_THRESHOLD, course.certificateDistinctionThreshold)
        values.put(DbStructureCourse.Columns.CERTIFICATE_REGULAR_THRESHOLD, course.certificateRegularThreshold)
        values.put(DbStructureCourse.Columns.CERTIFICATE_LINK, course.certificateLink)
        values.put(DbStructureCourse.Columns.IS_CERTIFICATE_AUTO_ISSUED, course.isCertificateAutoIssued)
        values.put(DbStructureCourse.Columns.IS_CERTIFICATE_ISSUED, course.isCertificateIssued)
        values.put(DbStructureCourse.Columns.LAST_DEADLINE, course.lastDeadline)
        values.put(DbStructureCourse.Columns.BEGIN_DATE, course.beginDate)
        values.put(DbStructureCourse.Columns.END_DATE, course.endDate)
        values.put(DbStructureCourse.Columns.SLUG, course.slug)

        values.put(DbStructureCourse.Columns.SCHEDULE_LINK, course.scheduleLink)
        values.put(DbStructureCourse.Columns.SCHEDULE_LONG_LINK, course.scheduleLongLink)
        values.put(DbStructureCourse.Columns.SCHEDULE_TYPE, course.scheduleType)

        values.put(DbStructureCourse.Columns.LAST_STEP, course.lastStepId)
        values.put(DbStructureCourse.Columns.LEARNERS_COUNT, course.learnersCount)
        values.put(DbStructureCourse.Columns.REVIEW_SUMMARY, course.reviewSummary)
        values.put(DbStructureCourse.Columns.TIME_TO_COMPLETE, course.timeToComplete)
        values.put(DbStructureCourse.Columns.OPTIONS, course.courseOptions?.let(gson::toJson))

        values.put(DbStructureCourse.Columns.IS_PAID, course.isPaid)
        values.put(DbStructureCourse.Columns.PRICE, course.price)
        values.put(DbStructureCourse.Columns.CURRENCY_CODE, course.currencyCode)
        values.put(DbStructureCourse.Columns.DISPLAY_PRICE, course.displayPrice)
        values.put(DbStructureCourse.Columns.PRICE_TIER, course.priceTier)

        return values
    }

    override fun populateNestedObjects(course: Course): Course =
        course.apply {
            introVideo = videoDao.get(course.introVideo?.id ?: -1) // less overhead vs immutability
        }

    override fun storeNestedObjects(persistentObject: Course) {
        persistentObject.introVideo?.let(videoDao::replace)
    }
}
